### PR TITLE
Feat/meta 3047 classification prop optimisation

### PR DESCRIPTION
**Pull request by BurakAltas**
## Change description

> Finalized code fixes/introduces:

- Fixes overflow issue on counter for performance metric
- Adds many additional logs including direct search to ES and detailed task insights
- Adds many new metrics to debug/trace easily
- Fixes typo perfromDirectIndexQuery -> performDirectIndexQuery
- Small adjustments made to addTagPropagation for optimization (double checking etc.. is optimized)
- Solves Janusgraph error that is thrown sometimes when an entity tried to be retrieved
- Optimizes the time greatly
- Solves task restart problem 
- Introduces new approach on notification that supports inline functionality

## Type of change
- [ ] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 1]
```diff
@@ -446,7 +446,8 @@ private void addTagPropagation(AtlasVertex fromVertex, AtlasVertex toVertex, Atl
     }
 
     public List<AtlasVertex> addTagPropagation(AtlasVertex classificationVertex, List<AtlasVertex> propagatedEntityVertices) throws AtlasBaseException {
-        List<AtlasVertex> ret = null;
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("addTagPropagation");
+        List<AtlasVertex> ret = new ArrayList<>();
```
Comment (DillionTan) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:450]: i think its better if we don't change the allocation logic, in the old way we only create the new arraylist when we want to add

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 2]
```diff
@@ -483,21 +486,11 @@ public List<AtlasVertex> addTagPropagation(AtlasVertex classificationVertex, Lis
...
-                if (existingEdge != null) {
-                    continue;
-                }
-
                 if (LOG.isDebugEnabled()) {
                     LOG.debug(" --> Adding propagated classification: [{}] to {} ({}) using edge label: [{}]", classificationName, getTypeName(propagatedEntityVertex),
                             GraphHelper.getGuid(propagatedEntityVertex), CLASSIFICATION_LABEL);
                 }
 
-                if (ret == null) {
```
Comment (DillionTan) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:497]: ref. above comment

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 3]
```diff
@@ -509,9 +502,10 @@ public List<AtlasVertex> addTagPropagation(AtlasVertex classificationVertex, Lis
                 AtlasClassification classification = entityRetriever.toAtlasClassification(classificationVertex);
 
                 context.recordAddedPropagation(entityGuid, classification);
+                RequestContext.get().endMetricRecord(countMetricRecorder);
             }
         }
-
+        if (ret.isEmpty()) return null;
```
Comment (DillionTan) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:508]: ref. above comment

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java (hunk 1 above)
Comment (DillionTan) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:450]: but not important if addTagPropagation is not called very frequently

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 4]
```diff
@@ -2748,28 +2751,66 @@ public List<String> propagateClassification(String entityGuid, String classifica
...
+                if (offset + CHUNK_SIZE > impactedVerticesSize) {
+                    toIndex = (int) impactedVerticesSize;
+                }
+                else {
+                    toIndex = offset + CHUNK_SIZE;
+                }
+                List<String> chunkedGuids = processChunkedPropagation(impactedVertices.subList(offset, toIndex), classificationVertex);
 
-            List<AtlasEntity> propagatedEntities = updateClassificationText(classification, entitiesPropagatedTo);
+                if((chunkedGuids != null) && (! chunkedGuids.isEmpty())){
```
Comment (DillionTan) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:2771]: is it possible that this is null here because addTagPropagation or some other logic returned null without throwing exception? in that case this logic will continue chunking even though one of the chunks returned null

File: repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java [hunk 5]
```diff
@@ -225,6 +224,10 @@ public void complete(AtlasVertex taskVertex, AtlasTask task) {
         }
 
         updateStatus(taskVertex, task);
+
+        graph.commit();
```
Comment (DillionTan) [inline repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java:228]: why is this needed?

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java (hunk 1 above)
Comment (BurakAltas) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:450]: Actually, if we were considering the heap usage, the size of an empty list is zero.

File: repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java (hunk 5 above)
Comment (BurakAltas) [reply repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java:228]: This was just making sure graph is committed at the end but it is invoked in later operations as well. I will delete this one.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java (hunk 4 above)
Comment (BurakAltas) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:2771]: Due to the line 2797, It can return null.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 6]
```diff
@@ -505,7 +505,6 @@ public List<AtlasVertex> addTagPropagation(AtlasVertex classificationVertex, Lis
                 RequestContext.get().endMetricRecord(countMetricRecorder);
             }
         }
-        if (ret.isEmpty()) return null;
```
Comment (DillionTan) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java]: even we remove null return here, line 2796 in EntityGraphMapper will return null for processChunkedPropagation if the collection is empty. Why not keep the null logic and check for null to exit early in propagateClassification?

my preference is still to initialize as null and only allocate new ArrayList when necessary

Comment (DillionTan) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java]: even we remove null return here, line 2796 in EntityGraphMapper will return null for processChunkedPropagation if the collection is empty. Why not keep the null logic and check for null to exit early in propagateClassification?

my preference is still to initialize as null and only allocate new ArrayList when necessary

Comment (BurakAltas) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java]: I do prefer this way instead of checking for if we should initialize the array or not for each vertex during propagation. Quite small heap usage which GC clears often vs if check in each propagation? It is also your decision of course I can change it easily. This is quite a small difference I believe so whichever way we choose I'm ok with.

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java [hunk 7]
```diff
@@ -144,6 +144,19 @@ private String perfromDirectIndexQuery(String query) throws IOException {
         return EntityUtils.toString(response.getEntity());
     }
 
+    private String performDirectIndexQueryWithSource(String query) throws IOException {
+        HttpEntity entity = new NStringEntity(query, ContentType.APPLICATION_JSON);
+
+        String endPoint = index + "/_search?_source=true";
```
Comment (mehtaanshul) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java:150]: Why do we need _source=true?

and if we need it, no need to create a separate function with a new name. You can use method overloading and call one method from the other to avoid code duplication.

Comment (BurakAltas) [reply graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java:150]: Because the task information is in the source of the document in Elasticsearch.

Will updated shortly.

Comment (BurakAltas) [reply graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java]: Can you please check [latest commit](https://github.com/atlanhq/atlas-metastore/pull/727/commits/2c2e6d69ee30291063d80f1292b1aae7d2e0ff20).

File: webapp/src/main/java/org/apache/atlas/notification/EntityNotificationListenerV2.java [hunk 8]
```diff
@@ -191,6 +193,27 @@ private void notifyClassificationEvents(List<AtlasEntity> entities, OperationTyp
         RequestContext.get().endMetricRecord(metric);
     }
 
+    private void notifyClassificationEvents(List<AtlasEntity> entities, OperationType operationType, Object mutatedObj, boolean forceInline) throws AtlasBaseException {
```
Comment (mehtaanshul) [inline webapp/src/main/java/org/apache/atlas/notification/EntityNotificationListenerV2.java:196]: Call this method in this method "private void notifyClassificationEvents(List<AtlasEntity> entities, OperationType operationType, Object mutatedObj)" here

File: repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java [hunk 9]
```diff
@@ -303,6 +304,10 @@ public List<AtlasTask> getTasksForReQueue() {
             ret = getTasksForReQueueIndexSearch();
         }
 
+        if (ret.size() > 0) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java:307]: I think we can remove this log

File: repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java [hunk 10]
```diff
@@ -380,7 +390,12 @@ public List<AtlasTask> getTasksForReQueueIndexSearch() {
                         AtlasVertex vertex = iterator.next().getVertex();
 
                         if (vertex != null) {
-                            ret.add(toAtlasTask(vertex));
+                            AtlasTask atlasTask = toAtlasTask(vertex);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java:393]: Want to understand this once

File: repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java [hunk 11]
```diff
@@ -210,7 +210,6 @@ public void inProgress(AtlasVertex taskVertex, AtlasTask task) {
         graph.commit();
     }
 
-    @GraphTransaction
     public void complete(AtlasVertex taskVertex, AtlasTask task) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java:213]: We have removed this for all kinds of tasks, will this affect other tasks?

File: webapp/src/main/java/org/apache/atlas/notification/EntityNotificationListenerV2.java [hunk 12]
```diff
@@ -98,7 +100,7 @@ public void onClassificationsAdded(AtlasEntity entity, List<AtlasClassification>
 
     @Override
     public void onClassificationsAdded(List<AtlasEntity> entities, List<AtlasClassification> classifications) throws AtlasBaseException {
-        notifyClassificationEvents(entities, CLASSIFICATION_ADD, classifications);
+        notifyClassificationEvents(entities, CLASSIFICATION_ADD, classifications, true);
```
Comment (mehtaanshul) [inline webapp/src/main/java/org/apache/atlas/notification/EntityNotificationListenerV2.java:103]: This won't just send inline for classification prop, but whenever classification gets added. Want to discuss this

**Review by mehtaanshul — COMMENTED** [review]
Added few comments

File: repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java (hunk 10 above)
Comment (BurakAltas) [reply repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java:393]: I've assigned it to a new AtlasTask object just for logging purposes

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java (hunk 7 above)
Comment (mehtaanshul) [reply graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java]: Discuss over call, Burak will make the changes

File: repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java (hunk 11 above)
Comment (mehtaanshul) [reply repository/src/main/java/org/apache/atlas/tasks/TaskRegistry.java:213]: Add it back based on discussion, and verify there's no negative impact

File: webapp/src/main/java/org/apache/atlas/notification/EntityNotificationListenerV2.java (hunk 12 above)
Comment (mehtaanshul) [reply webapp/src/main/java/org/apache/atlas/notification/EntityNotificationListenerV2.java:103]: Let's do this only for tasks

File: graphdb/api/src/main/java/org/apache/atlas/repository/graphdb/AtlasIndexQuery.java [hunk 13]
```diff
@@ -69,6 +69,13 @@
      */
     Long vertexTotals();
 
+    /**
+     * Directly queries Elasticsearch.
+     * @param searchParams
+     * @return result as String
+     */
+    String directElasticsearchQuery(SearchParams searchParams);
```
Comment (mehtaanshul) [inline graphdb/api/src/main/java/org/apache/atlas/repository/graphdb/AtlasIndexQuery.java:77]: remove this

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java [hunk 14]
```diff
@@ -198,6 +198,15 @@ public Long vertexTotals() {
         return vertexTotals;
     }
 
+    @Override
+    public String directElasticsearchQuery(SearchParams searchParams) {
```
Comment (mehtaanshul) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java:202]: remove this

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasJanusIndexQuery.java [hunk 15]
```diff
@@ -114,6 +114,11 @@ public Long vertexTotals() {
         return query.vertexTotals();
     }
 
+    @Override
```
Comment (mehtaanshul) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasJanusIndexQuery.java:117]: remove this

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java [hunk 16]
```diff
@@ -131,10 +131,10 @@ private DirectIndexQueryResult runQueryWithLowLevelClient(SearchParams searchPar
         return result;
     }
 
-    private String performDirectIndexQuery(String query, boolean includeSource) throws IOException {
+    private String performDirectIndexQuery(String query) throws IOException {
```
Comment (mehtaanshul) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasElasticsearchQuery.java:134]: Since we are only using source=false as of now, let's not change this function

**Review by mehtaanshul — APPROVED** [review]

**Review by sumandas0 — APPROVED** [review]

**Review by mehtaanshul — APPROVED** [review]
